Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h

```diff
@@ -93,126 +91,165 @@ class Triangulate_modifier
     return fh->info().is_external;
   }
 
-  void triangulate_face(face_descriptor f, PM& pmesh)
+  bool triangulate_face(face_descriptor f, PM& pmesh)
   {
-    typename Traits::Vector_3 normal =
-      Polygon_mesh_processing::compute_face_normal(f, pmesh);
 
-    P_traits cdt_traits(normal);
-    CDT cdt(cdt_traits);
 
-    // Halfedge_around_facet_circulator
-    typedef typename CDT::Vertex_handle Tr_Vertex_handle;
-    halfedge_descriptor start = halfedge(f, pmesh);
-    halfedge_descriptor h = start;
-    Tr_Vertex_handle previous, first;
-    do
+    typename Traits::Vector_3 normal =
+      Polygon_mesh_processing::compute_face_normal(f, pmesh);
+    if(normal == typename Traits::Vector_3(0,0,0))
+      return false;
+    std::size_t original_size = CGAL::halfedges_around_face(halfedge(f, pmesh), pmesh).size();
+    if(original_size == 4)
     {
-      Tr_Vertex_handle vh = cdt.insert(_vpmap[target(h, pmesh)]);
-      if (first == Tr_Vertex_handle()) {
-        first = vh;
+      typename Kernel::Point_3 p0, p1, p2, p3;
+      halfedge_descriptor v0, v1, v2, v3;
+      v0 = halfedge(f, pmesh);
+      p0 = _vpmap[target(v0, pmesh)];
+      v1 = next(v0, pmesh);
+      p1 = _vpmap[target(v1, pmesh)];
+      v2 = next(v1, pmesh);
+      p2 = _vpmap[target(v2, pmesh)];
+      v3 = next(v2, pmesh);
+      p3 = _vpmap[target(v3, pmesh)];
+
+      bool predicate = CGAL::cross_product(p2-p1,p0-p1) * CGAL::cross_product(p0-p3,p2-p3) > 0;
```

Comment:
That is not a predicate in the sense of CGAL.

I have the impression that the functor [`Kernel::Compare_dihedral_angle_3`][doc] from the kernel could be used, with `0` for the cosine.

[doc]: http://doc.cgal.org/latest/Kernel_23/classKernel.html#a00792fd1a5d950d54a750f4aa9d2ba3f

But actually, to be efficient, a dedicated predicate would be better. `Compare_dihedral_angle_3` computes cross products `u` and `v`, and their scalar product `u*v`. But it also computes the squared length of `u` and `v`: it checks `(u*v)^2 > cosine * ((u*u)*(v*v))`. The computations of lengths are useless when one want to compare the sign of `u*v` with 0.